sled-agent/src/opte.rs

```diff
@@ -0,0 +1,334 @@
+                    port_name: name.clone(),
+                    dest: cidr,
+                    target: RouterTarget::VpcSubnet(IpCidr::Ip4(cidr)),
+                };
+                hdl.add_router_entry_ip4(&route)?;
+            }
+            IpAddr::V6(_) => {
+                return Err(opte_ioctl::Error::InvalidArgument(String::from(
+                    "IPv6 not yet supported",
+                ))
+                .into());
+            }
+        }
+
+        // Create a VNIC on top of this device, to hook Viona into.
+        //
+        // Viona is the virtio-net driver used to provide virtual interfaces to
+        // guests. It sits on top of a MAC provider, which is responsible for
+        // delivering frames to the underlying data link. In theory, Viona can
+        // work with any MAC provider. However, there are several implicit
+        // assumptions, in both Viona _and_ MAC, that require Viona to be built
+        // on top of a VNIC specifically. There is probably a good deal of work
+        // required to relax that assumption, so in the meantime, we create a
+        // superfluous VNIC on the OPTE device, solely so Viona can use it.
+        let vnic = {
+            let phys = PhysicalLink(name.clone());
+            let vnic_name = format!("v{}", name);
+            Dladm::create_vnic(
+                &phys,
+                &vnic_name,
+                Some(omicron_common::api::external::MacAddr(mac)),
+                None,
+            )?;
+            Some(Vnic::wrap_existing(vnic_name))
+        };
+
+        Ok(OptePort {
+            name,
+            ip,
+            subnet,
+            mac,
+            vni,
+            underlay_ip,
+            gateway,
+            boundary_services,
+            vnic,
+        })
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct BoundaryServices {
+    pub ip: Ipv6Addr,
+    pub vni: Vni,
+}
+
+impl Default for BoundaryServices {
+    fn default() -> Self {
+        // TODO-completeness: Don't hardcode this.
+        //
+        // Boundary Services will be started on several sleds during rack
+        // setup, and the addresses of the chosen sleds will need to be
+        // propagated here.
+        const BOUNDARY_SERVICES_ADDR: Ipv6Addr =
+            Ipv6Addr::new(0xfd00, 0x99, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01);
+        let boundary_services_vni = Vni::new(99_u32).unwrap();
+
+        Self { ip: BOUNDARY_SERVICES_ADDR, vni: boundary_services_vni }
+    }
+}
+
+/// Information about the gateway for an OPTE port
+#[derive(Debug, Clone, Copy)]
+pub struct Gateway {
+    mac: MacAddr6,
+    ip: IpAddr,
+}
+
+impl Gateway {
+    pub fn from_subnet(subnet: &IpNetwork) -> Self {
+        Self {
+            // TODO-completeness: We need a robust way of assigning a
+            // non-clashing MAC address for the virtual gateway that OPTE
+            // represents.
+            mac: MacAddr6::new(0xa8, 0x40, 0x25, 0x00, 0x00, 0x01),
```

Comment:
Note to myself: Whatever we pick here needs to be excluded as an option in the Nexus code that creates random MAC addresses for these.

Reply:
@rzezeski Are you sure it shouldn't be `FF:77:77`, you know, to make sure you get your Final Fantasy in there.

More seriously for the moment, I'm not sure how we want to be generally allocating MACs. Because we're using `/32` or `/128` instances here, the actual MAC address technically doesn't have to be unique between instance (though they must be unique within an instance). Separately, though, we generally have a limit on the number of instances that are within a single VPC / project. While I don't know that we have this right now (do we @davepacheco?), it's worth teasing this apart.

Today we have a very simple model where each gateway is basically one 'hop' so to say to the next place. And it can be the same MAC for everyone, because well, that's fine. However, let's imagine a world where we want thinks to be unique for some reason. That is, would it make sense to further subdivide this space in Nexus for future us in case we change? Is subdividing space even the right approach?

There is a surprising amount of nuance here. There are a few different things I think about when we do this:

1. How do we actually track in our alloactor what is allocated and what is free?
2. Do we want to programmatically make decisions based on a range or is there something useful for humans when debugging?
3. What is the cost of splitting the space and does it create arbitrary limits that we don't want?

I have less clarify on how we're doing (1), but my current assumption based on the comments here is that we're basically generating a random ID and then verifying it's not in use and if it is, taking another lap. Note, as the space fills up (e.g. at least greater than 50% threshold probably, but I'm guessing and not mathematically deriving this), we probably want to do random and then scan for the next unused one

On (2), the answer is honestly I don't know. There is probably value in debugging for us as humans at some point in being able to do pattern recognition. However, I'm not sure what we expect from a programatic perspective. If I think about the implementation of OPTE for a moment, because it's all `/32` and `/128` addresses in VMs, it's a little inconsequential and really it's just matching on explicit things as opposed to classes (right now). If we ever end up in a world where we implement something ala Hoverboards, it may become more useful depending on how we want to phrase things in the packet itself and track its journey through different endpoints. This is also in my head for when we have cross-AZ style activity or similar. Further, because of our current strategy for (1), splitting becomes somewhat more appealing.

With that in mind, I think it may be OK to further subdivide our space of virtual IPs. This should get codifid in a new RFD which is basically around how the control plane does all its networking, e.g. implementation details of some of the bootstrap, this, the VPC ID discussion, some recent RFD 63 bits, etc.

With this I would further propose breaking things down as:

* `F0:00:00` - `FE:FF:FF`: Use for customer instances and related.  Results in a fair number of IDs, e.g. ~983040.
* `FF:00:00` - `FF:FF:FF`: Reserve for VPC system MAC addresses. Results in about 64K IDs (I think).

Today, we basically would then use the static allocation for all gateways right now. I'd like to honor @rzezeski's MAC request given all the OPTE work, if possible. Right now we would ignore any allocations of the VPC / system addresses. Those may later be used also for other things like VPC gateways, dealing with IPSec, etc.

Whether we want to go revisit the uniqueness of the MACs in the instances is also up to us. I think it depends on basically whether we would ever want to offer up L2 broadcast/multicast in the future, in which case we'd want it to be unique within at least the VPC Subnet. Anyways, there'll be more on bits of this as I answer the other higher-level questions here. Note, none of this should block the current stuff.

Reply:
This all sounds reasonable, thanks for the detailed write-up Robert. We can definitely honor @rzezeski's request, and split the virtual MAC space further into "customer" and "system" addresses without too much fuss.

Your guess is mostly correct @rmustacc that we're randomly assigning MAC addresses within the space of virtual addresses, but we're _not_ currently retrying anything. I believe the only place we currently retry automatically is when generating a random IPv6 subnet for a VPC Subnet, assuming one isn't provided by the client. In this case, if there's a conflict, we return a 500 to the caller. That sucks, but I think the work to retry just got deferred.

I bring that up because I'm just trying to be clear about what I think is reasonable for this PR. Splitting the MAC address space and providing `A8:40:25:FF:77:77` for OPTE is in scope. Doing any sort of automatic retry when that random assignment conflicts is out of scope, I would say. (I will add an issue to track.) Doing a linear-probe kind of allocation, i.e. the random + scan approach, is a lot of work too in my experience with CockroachDB, so out of scope.

One thing I'm less clear on is the scope within which MACs are unique. It's currently VPC. Making it VPC Subnet would be a one-line change to the database index. But it sounds like there are more significant questions to write up and wrestle with about this. Is that change valuable now? Anything more, I'd like to defer, but I can also just not do anything in this PR.

Reply:
My goal is not to try and get this all done in this PR. I think it's very reasonable to use the current allocation policy around MACs, change the split / MAC, and then come back to the rest of this in a subsequent thing. As I've mentioned, I'd like us to write up an RFD on this and management.

It's fine to do all the retry logic in a follow up, but it's something we'll want to get nailed before we ship.

Reply:
Sounds great, thanks for clarifying. I'll update the MAC address space and assign what shall henceforth be known as "rpz's MAC" here, and then create issues to track the retry and possibly smarter allocation. Thanks!

Reply:
Tracking the retry part of this in https://github.com/oxidecomputer/omicron/issues/963. I started making an issue around "smarter allocation," but it feels more like an RFD than an issue. Or at least, I don't want to make an issue that has so little actionable relationship to the code. I'll think more about that.